src/test/java/option/OptionalExample.java

```diff
@@ -4,12 +4,42 @@
 
 import java.util.Optional;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.BiFunction;
 import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
 
 import static org.junit.Assert.assertEquals;
 
 public class OptionalExample {
 
+    public static <T1, T2, R> Optional<R> zipMap1(Optional<T1> o1, Optional<T2> o2, BiFunction<T1, T2, R> f) {
+        if (o1.isPresent() && o2.isPresent()) {
+            return Optional.ofNullable(f.apply(o1.get(), o2.get()));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    public static <T1, T2, R> Optional<R> zipMap2(Optional<T1> o1, Optional<T2> o2, BiFunction<T1, T2, R> f) {
+        return o1.flatMap(
+                t1 -> o2.flatMap(
```

Comment:
You can use `map` here. `option.map(f)` returns empty option in case of `null` result of `f`.